### PR TITLE
[Dam] Fix construction utility

### PR DESCRIPTION
**Pull request by joaquinirazabal**
**📝 Description**
This PR fixes the segmentation fault problem reported in #10877.

Please mark the PR with appropriate tags: 
- Bugfix

**🆕 Changelog**
- Fix problem with the initilization of elements when using the construction utility of the DamApplication
- Transform some OMP loops into parallel utilities format
- Add test for the construction utility of the DamApplicatin

### CODE REVIEW

File: applications/DamApplication/custom_utilities/construction_utility.hpp [hunk 1]
```diff
@@ -88,99 +88,75 @@ class ConstructionUtility
...
             // Same nodes for both computing model part
-            ModelPart::NodesContainerType::iterator it_begin = mrThermalModelPart.NodesBegin();
-            #pragma omp parallel for
-            for (int i = 0; i < nnodes; ++i)
-            {
-                ModelPart::NodesContainerType::iterator it = it_begin + i;
-                it->Set(ACTIVE, false);
-                it->Set(SOLID, false);
-            }
+            block_for_each(mrThermalModelPart.Nodes(), [&](auto& rNode){
```
Comment (philbucher) [inline applications/DamApplication/custom_utilities/construction_utility.hpp]: awesome!

File: applications/DamApplication/custom_utilities/construction_utility.hpp [hunk 2]
```diff
@@ -88,99 +88,75 @@ class ConstructionUtility
...
-            const int soil_nnodes = mrMechanicalModelPart.GetSubModelPart(mMechanicalSoilPart).Nodes().size();
 
             if (soil_nelements != 0)
             {
-                ModelPart::ElementsContainerType::iterator el_begin = mrMechanicalModelPart.GetSubModelPart(mMechanicalSoilPart).ElementsBegin();
-                ModelPart::ElementsContainerType::iterator el_begin_thermal = mrThermalModelPart.GetSubModelPart(mThermalSoilPart).ElementsBegin();
-                mNumNode = el_begin->GetGeometry().PointsNumber();
+                block_for_each(mrMechanicalModelPart.GetSubModelPart(mMechanicalSoilPart).Elements(), [&](auto& rMechanicalElement){
+                    rMechanicalElement.Set(ACTIVE, true);
+                });
```
Comment (philbucher) [inline applications/DamApplication/custom_utilities/construction_utility.hpp]: I think this function already exists in the VariableUtils

File: applications/DamApplication/custom_utilities/construction_utility.hpp [hunk 3]
```diff
@@ -241,24 +217,16 @@ class ConstructionUtility
...
-                ModelPart::NodesContainerType::iterator it_begin = mrThermalModelPart.NodesBegin();
-                #pragma omp parallel for
-                for (int i = 0; i < nnodes; ++i)
-                {
-                    ModelPart::NodesContainerType::iterator it = it_begin + i;
-                    it->FastGetSolutionStepValue(ALPHA_HEAT_SOURCE) = mAlphaInitial;
-                }
+                block_for_each(mrThermalModelPart.Nodes(), [&](auto& rNode){
+                    rNode.FastGetSolutionStepValue(ALPHA_HEAT_SOURCE) = mAlphaInitial;
+                });
```
Comment (philbucher) [inline applications/DamApplication/custom_utilities/construction_utility.hpp]: same

File: applications/DamApplication/tests/construction/construction.mdpa
Comment (philbucher) [inline applications/DamApplication/tests/construction/construction.mdpa:1]: We try to keep mdpas below 1000 lines, to have fast tests and keep the size of the repo in bounds

File: applications/DamApplication/custom_utilities/construction_utility.hpp (hunk 2 above)
Comment (joaquinirazabal) [reply applications/DamApplication/custom_utilities/construction_utility.hpp]: Great advice, thank you! :)

File: applications/DamApplication/custom_utilities/construction_utility.hpp (hunk 1 above)
Comment (joaquinirazabal) [reply applications/DamApplication/custom_utilities/construction_utility.hpp]: Thanks ;)

File: applications/DamApplication/tests/construction/construction.mdpa
Comment (joaquinirazabal) [reply applications/DamApplication/tests/construction/construction.mdpa:1]: I reduced the mdpa as much as I could

**Review by philbucher — APPROVED** [review]

**Pull request merged**